es6-shim.js

```diff
@@ -2097,6 +2170,257 @@
+          if (!ES.TypeIsObject(target)) {
+            throw new TypeError('target must be an object');
+          }
+
+          var receiver = arguments.length > 2 ? arguments[2] : target;
+
+          return internal_get(target, key, receiver);
+        },
+
+        getOwnPropertyDescriptor: throwUnlessTargetIsObject(Object.getOwnPropertyDescriptor),
+
+        getPrototypeOf: throwUnlessTargetIsObject(Object.getPrototypeOf),
+
+        has: function has(target, key) {
+          if (!ES.TypeIsObject(target)) {
+            throw new TypeError('target must be an object');
+          }
+
+          return key in target;
+        },
+
+        isExtensible: throwUnlessTargetIsObject(Object.isExtensible),
+
+        // Basically the result of calling the internal [[OwnPropertyKeys]].
+        // Concatenating propertyNames and propertySymbols should do the trick.
+        // This should continue to work together with a Symbol shim
+        // which overrides Object.getOwnPropertyNames and implements
+        // Object.getOwnPropertySymbols.
+        ownKeys: function ownKeys(target) {
+          if (!ES.TypeIsObject(target)) {
+            throw new TypeError('target must be an object');
+          }
+
+          var keys = Object.getOwnPropertyNames(target);
+
+          if (ES.IsCallable(Object.getOwnPropertySymbols)) {
+            keys = keys.concat(Object.getOwnPropertySymbols(target));
+          }
+
+          return keys;
+        },
+
+        preventExtensions: wrapObjectFunction(Object.preventExtensions),
+
+        set: function set(target, key, value) {
+          if (!ES.TypeIsObject(target)) {
+            throw new TypeError('target must be an object');
+          }
+
+          var receiver = arguments.length > 2 ? arguments[2] : target;
+
+          return internal_set(target, key, value, receiver);
+        },
+
+        // Sets the prototype of the given object.
+        // Returns true on success, otherwise false.
+        setPrototypeOf: function setPrototypeOf(object, proto) {
+          if (!ES.TypeIsObject(object)) {
+            throw new TypeError('target must be an object');
+          }
+
+          if (proto !== null && !ES.TypeIsObject(proto)) {
+            throw new TypeError('proto must be an object or null');
+          }
+
+          // If they already are the same, we're done.
+          if (proto === Reflect.getPrototypeOf(object)) {
+            return true;
+          }
+
+          // Cannot alter prototype if object not extensible.
+          if (!Reflect.isExtensible(object)) {
+            return false;
+          }
+
+          // Ensure that we do not create a circular prototype chain.
+          if (proto !== null) {
+            var p = proto;
+
+            while (p) {
+              if (object === p) {
+                return false;
+              }
+
+              p = Reflect.getPrototypeOf(p);
+            }
+          }
+
+          Object.setPrototypeOf(object, proto);
```

Comment:
What happens in a browser where `Object.setPrototypeOf` is not defined (because it sometimes can't be shimmed)? I'd think `Reflect.setPrototypeOf` should similarly not be shimmed in the first place, in this case.